server/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java

```diff
@@ -246,10 +258,132 @@ protected void reroute(ShardId shardId, String reason) {
+                                @Override
+                                public void onResponse(NodesGroupedGatewayStartedShards nodesGroupedStartedShards) {
+                                    assert nodesGroupedStartedShards.getNodes().size() == 1;
+                                    NodeGroupedGatewayStartedShards startedShardResponse = nodesGroupedStartedShards.getNodes().get(0);
+                                    List<NodeGatewayStartedShards> startedShards = startedShardResponse.getStartedShards();
+                                    for (NodeGatewayStartedShards startedShard : startedShards) {
+                                        // transfer to single shard -> nodes (single node) response
+                                        List<NodeGatewayStartedShards> listSingleStartedShards = new ArrayList<>();
+                                        listSingleStartedShards.add(startedShard);
+                                        NodesGatewayStartedShards newNodesResponse = new NodesGatewayStartedShards(
+                                            nodesGroupedStartedShards.getClusterName(),
+                                            listSingleStartedShards,
+                                            nodesGroupedStartedShards.failures());
+
+                                        // call child listener
+                                        Map<String, ActionListener<NodesGatewayStartedShards>> listeners =
+                                            queuedListeners.get(startedShard.getShardId());
+                                        ActionListener<NodesGatewayStartedShards> listener = listeners.get(node.getId());
+                                        listener.onResponse(newNodesResponse);
+
+                                        // clean child listener
+                                        ActionListener<NodesGatewayStartedShards> removed = listeners.remove(node.getId());
+                                        assert removed != null;
+                                        if (listeners.isEmpty()) {
+                                            queuedListeners.remove(startedShard.getShardId());
+                                        }
+                                    }
+                                }
+
+                                @Override
+                                public void onFailure(Exception e) {
+                                    // clean requests and listeners of this node
+                                    for (ShardId shardId : curNodeRequests.keySet()) {
+                                        curNodeRequests.get(shardId).getListener().onFailure(e);
+                                        queuedListeners.get(shardId).remove(node.getId());
+                                    }
+                                }
+                            };
+
+                            client.executeLocally(
+                                TransportNodesListGatewayStartedShards.TYPE,
```

Comment:
I'm undecided about re-using the same action type for both kinds of request here. I think it'd be cleaner to introduce a new one (and to name it something better than `internal:gateway/local/started_shards`) given how big a difference in behaviour we are making.

Reply:
Hi @DaveCTurner , if we introduce a new action, then we need to refactor some logics in `GatewayAllocator`, like the follow structures, it seems that would a big change for the high level allocators. How do you think so?
https://github.com/elastic/elasticsearch/blob/2629c32efd53809f1792f093567e19cfcf44bf29/server/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java#L57-L60

Reply:
I'm not sure this is true, I think we could keep pretty much the same interface from the point of view of `GatewayAllocator`. It should be possible to implement a batching `Lister` which reworks the batched responses into a `BaseNodesResponse<NodeGatewayStartedShards>`.